test/e2e/federated-service.go

```diff
@@ -0,0 +1,283 @@
+		}
+	})
+
+	AfterEach(func() {
+		framework.SkipUnlessFederated(f.Client)
+
+		// Delete the registered clusters in the federation API server.
+		clusterList, err := f.FederationClientset.Federation().Clusters().List(api.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		for _, cluster := range clusterList.Items {
+			err := f.FederationClientset.Federation().Clusters().Delete(cluster.Name, &api.DeleteOptions{})
+			Expect(err).NotTo(HaveOccurred())
+		}
+	})
+
+	It("should be able to discover a federated service", func() {
+		framework.SkipUnlessFederated(f.Client)
+
+		createService(f.FederationClientset, clusterClientSets, f.Namespace.Name)
+
+		svcDNSNames := []string{
+			FederatedServiceName,
+			fmt.Sprintf("%s.%s", FederatedServiceName, f.Namespace.Name),
+			fmt.Sprintf("%s.%s.svc.cluster.local.", FederatedServiceName, f.Namespace.Name),
+			fmt.Sprintf("%s.%s.%s", FederatedServiceName, f.Namespace.Name, federationName),
+			fmt.Sprintf("%s.%s.%s.svc.cluster.local.", FederatedServiceName, f.Namespace.Name, federationName),
+		}
+		for _, name := range svcDNSNames {
+			discoverService(f, name, true)
+		}
+	})
+
+	It("should be able to discover a non-local federated service", func() {
+		framework.SkipUnlessFederated(f.Client)
+
+		createService(f.FederationClientset, clusterClientSets, f.Namespace.Name)
+
+		// Delete a federated service shard in the default e2e Kubernetes cluster.
+		err := f.Clientset_1_3.Core().Services(f.Namespace.Name).Delete(FederatedServiceName, &api.DeleteOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		waitForFederatedServiceShard(f.Clientset_1_3, f.Namespace.Name, nil, 0)
+
+		localSvcDNSNames := []string{
+			FederatedServiceName,
+			fmt.Sprintf("%s.%s", FederatedServiceName, f.Namespace.Name),
+			fmt.Sprintf("%s.%s.svc.cluster.local.", FederatedServiceName, f.Namespace.Name),
+		}
+		for _, name := range localSvcDNSNames {
+			discoverService(f, name, false)
+		}
+
+		svcDNSNames := []string{
+			fmt.Sprintf("%s.%s.%s", FederatedServiceName, f.Namespace.Name, federationName),
+			fmt.Sprintf("%s.%s.%s.svc.cluster.local.", FederatedServiceName, f.Namespace.Name, federationName),
+		}
+		for _, name := range svcDNSNames {
+			discoverService(f, name, true)
+		}
+	})
+})
+
+// waitForFederatedServiceShard waits until the number of shards of a given federated
+// service reaches the expected value, i.e. numSvcs in the given individual Kubernetes
+// cluster. If the shard count, i.e. numSvcs is expected to be at least one, then
+// it also checks if the first shard's name and spec matches that of the given service.
+func waitForFederatedServiceShard(cs *release_1_3.Clientset, namespace string, service *api.Service, numSvcs int) {
+	By("Fetching a federated service shard")
+	var clSvcList *v1.ServiceList
+	if err := wait.PollImmediate(framework.Poll, FederatedServiceTimeout, func() (bool, error) {
+		var err error
+		clSvcList, err = cs.Core().Services(namespace).List(api.ListOptions{})
```

Comment:
Don't you only want to list the service with the given name (service.Name()) here?  I realize that this test should run in it's own namespace, and that there should be no other services in that namespace, but the logic in this function still seems overly complex and confusing to me. 

Should you not simply be listing the services with the given name, and confirming that the length of the returned list is either 0 or 1 (depending on the value of numServices, which I suspect should be a bool, not an int).